dali/operators/decoder/video/video_decoder_base.h

```diff
@@ -15,121 +15,330 @@
+          },
+          output[s].shape().num_elements());
     }
-    return shape;
+    thread_pool.RunAll();
   }
 
+  ThreadPool &GetThreadPool(const Workspace &ws) {
+    return std::is_same<MixedBackend, Backend>::value ? *thread_pool_ : ws.GetThreadPool();
+  }
 
-  /**
-   * Decode given number of frames from a FramesDecoder.
-   *
-   * This function allows optional padding of the output data, when
-   * there's not enough frames inside FramesDecoder to fill the `num_frames` specified.
-   *
-   * The `pad_value` argument determines, with what the VideoDecoder shall pad
-   * the partial sequence. The tensor passed to the `pad_value` shall be an entire
-   * frame. This frame will be repeated at the end of partial sequence.
-   * If the `pad_value` is not provided, the padding will not happen and the DecodeFrames
-   * will return a partial sequence.
-   *
-   * @param output The SampleView in which the decoded sequence will be put.
-   * @param sample_idx Index of the encoded video that shall be decoded.
-   * @param num_frames How many frames shall be decoded.
-   * @param pad_value What to (optionally) pad the partial sequence with.
-   * @return False, if less than `num_frames` have been decoded.
-   */
-  bool DecodeFrames(SampleView<OutBackend> output, int64_t sample_idx, int64_t num_frames,
-                    std::optional<SampleView<OutBackend>> pad_value = std::nullopt,
-                    std::optional<cudaStream_t> stream = std::nullopt) {
+  void DecodeFrames(span<const FrameInfo> frame_infos, SampleView<OutBackend> output,
+                    int64_t sample_idx, cudaStream_t stream = 0) {
     auto &frames_decoder = *frames_decoders_[sample_idx];
     int64_t frame_size = frames_decoder.FrameSize();
+    uint8_t *output_data = output.template mutable_data<uint8_t>();
 
-    DALI_ENFORCE(!pad_value.has_value() || pad_value->shape().num_elements() == frame_size,
-                 make_string("Provided pad_value has improper number of elements. Expected: ",
-                             frame_size, "; Actual: ", pad_value->shape().num_elements()));
+    // Sort the frame indices so that we can decode them in natural order
+    // to avoid unnecessary seeking when possible
+    FrameInfo last_frame;
+    for (auto &frame : frame_infos) {
+      if (frame.is_constant) {
+        kernels::memset<OutStorage>(output_data + frame.frame_num * frame_size,
+                                    frame.constant.value, frame_size, stream);
+        continue;
+      } else if (frame.frame.frame_idx == last_frame.frame.frame_idx) {
+        kernels::copy<OutStorage, OutStorage>(output_data + frame.frame_num * frame_size,
+                                              output_data + last_frame.frame_num * frame_size,
+                                              frame_size, stream);
+        continue;
+      }
+      // Only seek if frame is not close to current position
+      int current_frame = frames_decoder.NextFrameIdx();
```

Comment:
I would add an API in the decoder that would tell us the index of the next key frame. If it is < that the next frame we are looking for we can seek, if not then there is not point as we will end up in the earlier frame than we are now.

Reply:
I will build this into the SeekFrame logic